src/lib/components/ScrollSpy/index.js

```diff
@@ -37,57 +43,28 @@ class ScrollSpy extends BaseElement {
     if (!this.articleContent) {
       return;
     }
-
-    this.headings = this.articleContent.querySelectorAll(
-      'h1[id], h2[id], h3[id]',
-    );
-
+    this.items = this.articleContent.querySelectorAll(this.selector);
     this.observer = new IntersectionObserver(this.scrollSpy, {
+      // Observe the upper parth of the viewport (user's focus area)
       rootMargin: '0px 0px -80% 0px',
     });
-    this.headings.forEach((heading) => this.observer.observe(heading));
+    this.items.forEach((item) => this.observer.observe(item));
   }
 
   disconnectedCallback() {
     super.disconnectedCallback();
     this.observer.disconnect();
   }
 
-  scrollSpy(headings) {
-    const links = new Map(
-      [...this.querySelectorAll('a')].map((l) => [l.getAttribute('href'), l]),
-    );
-    for (const heading of headings) {
-      const href = `#${heading.target.getAttribute('id')}`;
-      const link = links.get(href);
-
-      if (link) {
-        if (heading.intersectionRatio > 0) {
-          link.classList.add(this.tocVisibleClass);
-          this.previouslyActiveHeading = heading.target.getAttribute('id');
-        } else {
-          link.classList.remove(this.tocVisibleClass);
-        }
-      }
-
-      const firstVisibleLink = this.querySelector(`.${this.tocVisibleClass}`);
-
-      links.forEach((link) => {
-        link.classList.remove(this.tocActiveClass, this.tocVisibleClass);
-      });
-
-      if (firstVisibleLink) {
-        firstVisibleLink.classList.add(this.tocActiveClass);
-      }
-
-      if (!firstVisibleLink && this.previouslyActiveHeading) {
-        const last = this.querySelector(
-          `a[href="#${this.previouslyActiveHeading}"]`,
-        );
-        last.classList.add(this.tocActiveClass);
-      }
-    }
+  scrollSpy(items) {
+    const event = new CustomEvent('viewport-entry', {
```

Comment:
What ends up happening if you have, say, three different `ScrollSpy` instances on a given page, and each fires their own `'viewport-entry'` events? The code that listens for the events might only care about the ones fired from a specific `ScrollSpy` instance, but the listener callback will be invoked for all of them. This introduces a burden on the listener callback to filter out and ignore events that it doesn't care about.